### PR TITLE
sql: support binary format for all arrays

### LINKED ISSUE #21433: sql/pgwire: support ArrayOid decoding in pgwire binary format
Our pgwire implementation is currently able to decode many more array types in the text format than in the binary format by [treating them like strings](https://github.com/cockroachdb/cockroach/blob/40bed20826fc4f6e714d4be2ffad7400922f9b90/pkg/sql/pgwire/types.go#L669). This leads surprising results where certain types are only supported when a client driver uses the text format to send arrays of datums. An example of this is boolean arrays (oid 1000). Since applications often have little control over the format that drivers send datums in, we should make an effort to maintain parity between the two formats both for decoding and encoding.

See https://forum.cockroachlabs.com/t/bool-arrays-incompatible-with-github-com-lib-pq/1234/7

### LINKED ISSUE #23063: sql: unable to provide placeholder values of type uuid[] over pgwire
Reported on https://forum.cockroachlabs.com/t/sql-array-of-uuids-as-placeholder-for-select/1399

> The problem is with the following SQL query, that is generated by ecto:
>
>      SELECT u0.“url”, u0.“link_id”, u0.“link_id” FROM “urls” AS u0 
>       WHERE (u0.“link_id” = ANY($1)) ORDER BY u0.“link_id” 
>       [[<<161, 175, 147, 55, 25, 194, 69, 73, 149, 91, 215, 147, 152, 181, 29, 161>>, <<71, 153, 19, 83, 131, 79, 72, 21, 137, 151, 199, 190, 18, 60, 80, 99>>]]
>
> Note that this is the Elixir console output from Ecto. <<…>> is the Elixir representation for binaries.
> The following error is raised:
>
>       unsupported OID 2951 with format code formatBinary

The placeholder value for the single placeholder is really the expression `ARRAY[’{a1af9337-19c2-4549-955b-d79398b51da1}’, ‘{47991353-834f-4815-8997-c7be123c5063}’])`

### PR DESCRIPTION
**Pull request by justinj**
Fixes #23063.
Fixes #21433.

More testing probably needed but wanted to get this guy out there, and
see if anyone has any ideas for a nice way to do more large scale
testing than generating test data for every type of array we support?

Release note (bug fix): ARRAYs can now be used with the postgres binary
format.

### CODE REVIEW

**Comment by hjames9** [comment]
Would this work for creating an array type for jsonb columns by any chance?

**Review by knz — APPROVED** [review]
thanks!

plz cherry-pick too

File: pkg/cmd/generate-binary/main.go [hunk 1]
```diff
@@ -216,6 +217,12 @@ var inetInputs = []string{
 	"192.168./10",
 }
 
+var arrayUUIDInputs = []string{
```
Comment (nvanbenschoten) [inline pkg/cmd/generate-binary/main.go]: Consider adding a test for `decimalArrayInputs` as well.

File: pkg/sql/pgwire/types.go [hunk 2]
```diff
@@ -369,7 +369,8 @@ func (b *writeBuffer) writeBinaryDatum(
 		subWriter.putInt32(int32(hasNulls))
 		subWriter.putInt32(int32(v.ParamTyp.Oid()))
 		subWriter.putInt32(int32(v.Len()))
-		subWriter.putInt32(int32(v.Len()))
+		// Lower bound, we only support a lower bound of 1.
+		subWriter.putInt32(1)
```
Comment (nvanbenschoten) [inline pkg/sql/pgwire/types.go]: I'm curious how you caught this. What was failing?

**Review by nvb — APPROVED** [review]
Nice!

File: pkg/sql/pgwire/types.go (hunk 2 above)
Comment (justinj) [reply pkg/sql/pgwire/types.go]: The tests I just added were, haha. But I guess this is a good indicator that nobody really makes use of the lower bound.

File: pkg/cmd/generate-binary/main.go (hunk 1 above)
Comment (justinj) [reply pkg/cmd/generate-binary/main.go]: Good idea, I just took the tests for decimal and arbitrarily grouped them into arrays.

**Comment by justinj** [comment]
@hjames9 We punted arrays of `JSONB` values for the meantime because it causes some annoyances with respect to formatting. Out of curiosity - what do you want them for that you couldn't just use a JSONB-level array for?

**Comment by hjames9** [comment]
I have an existing Postgresql schema that stores json obtained from REST
calls in unrelated systems.  For instance, like a payment processing system
that calls one service to process a credit card and one to process
inventory moving.  The responses back from each system I would like to keep
separate and not try to comingle them into one giant message.

**Comment by knz** [comment]
What about storing the responses in multiple SQL rows instead of an array?
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by hjames9** [comment]
That's usually the suggestion for avoiding using array types, however I'd
prefer to use the same schema and not make too many code changes if
possible, especially considering that the intention is to support arrays in
general.

**Comment by justinj** [comment]
TFTRs!

**Pull request merged**